app/controllers/provider_foreman_controller.rb

```diff
@@ -208,26 +189,18 @@ def provider_foreman_form_fields
   end
 
   def authentication_validate
-    if params[:log_password]
-      if params[:provtype] == 'Ansible Tower'
-        @provider_cfgmgmt = ManageIQ::Providers::AnsibleTower::Provider.new(:name       => params[:name],
-                                                                            :url        => params[:url],
-                                                                            :zone_id    => Zone.find_by_name(MiqServer.my_zone).id,
-                                                                            :verify_ssl => params[:verify_ssl].eql?("on"))
-
-      else
-        @provider_cfgmgmt = ManageIQ::Providers::Foreman::Provider.new(:name       => params[:name],
-                                                                       :url        => params[:url],
-                                                                       :zone_id    => Zone.find_by_name(MiqServer.my_zone).id,
-                                                                       :verify_ssl => params[:verify_ssl].eql?("on"))
-      end
+    @provider_cfgmgmt = if params[:log_password]
+      Provider.new(
+        :name       => params[:name],
+        :url        => params[:url],
+        :type       => "ManageIQ::Providers::#{params[:provtype].delete("/\s+/")}::Provider",
```

Comment:
this is so much better.

I think a 2 line method `provider_type(params[:provtype])` would suffice and go a long way.